Add `repeat` to the specification

kgryte opened the pull request:
This PR

- resolves https://github.com/data-apis/array-api/issues/654 by adding support for `repeat` to the array API specification.
- allows `repeats` to be either an `int` or an `array`. NumPy and other inspired libraries and TensorFlow support one-dimensional arrays. NumPy also supports lists and tuples. CuPy docs suggest support for only lists and tuples. PyTorch supports a one-dimensional array; however, there has been discussion (linked to in the linked RFC) preferring sequences over arrays due to synchronization issues. However, it's not clear that providing a sequence of integers is particularly common or useful. In this PR, I've chosen to explicitly type `repeats` to support `int` and array. Should sequences be considered acceptable, this can be revisited in a future revision of the Array API standard.
- adds a data-dependent admonition and allows array libraries to omit array support for the ``repeats`` argument.

kgryte commented:
I've updated the proposed specification to include a note advising conforming array libraries to include a warning regarding device synchronization if `repeats` is an array.

kgryte commented:
@leofang Would you mind giving this PR a review? I believe this PR addresses the concerns you raised in https://github.com/data-apis/array-api/issues/654#issuecomment-1731612375, but I want to confirm before merging.

kgryte commented:
@leofang Pinging in case you missed the above.

leofang commented on src/array_api_stubs/_draft/manipulation_functions.py:
```diff
@@ -159,6 +160,51 @@ def permute_dims(x: array, /, axes: Tuple[int, ...]) -> array:
+        If ``repeats`` is an array, the array must have an integer data type.
+
+
+    .. note::
+       For specification-conforming array libraries supporting hardware acceleration, providing an array for ``repeats`` may cause device synchronization due to an unknown output shape. For those array libraries where synchronization concerns are applicable, conforming array libraries are advised to include a warning in their documentation regarding potential performance degradation when ``repeat ...
```
Q: Is this indentation level right?

kgryte replied:
Thanks for catching. Fixed!

leofang reviewed (approved):
Thanks for pushing this, Athan. Sorry I missed the ping. Took a stab at it, no concerns.

leofang reviewed (commented):
One more question: Should we also add a note on "data-dependent output shapes" like what we do for `unique*`/`nonzero`?

kgryte commented:
@leofang Re: admonition. I am not certain. It's only when `repeats` is an array that this is an issue correct?

And for that case, we include a note regarding device synchronization. If we add a "data-dependent admonition" here, this would make this API optional, which I am not sure is desirable.

As a point of reference, in `tile` (https://github.com/data-apis/array-api/pull/692), we did not include an admonition, even though `repetitions` can be a tuple of ints. If the `Sequence[int]` is problematic for `repeats`, I would imagine that we'd also need to add the admonition to `tile`.

leofang commented:
> Re: admonition. I am not certain. It's only when `repeats` is an array that this is an issue correct?

Doesn't this API always have the output shape determined by the input data (`repeats`), regardless of how `repeats` is provided?

rgommers commented:
> Doesn't this API always have the output shape determined by the input data (`repeats`), regardless of how `repeats` is provided?

Kinda sorta, but I think the "data-dependent shape" admonition is more aimed at the input values of the input array. E.g., the most common usage here will be with a literal int: `repeat(x, 2)`. Which isn't data-dependent.

I played with this a bit with JAX:
```python
>>> import jax
>>> import jax.numpy as jnp

>>> x = jnp.arange(3)
>>> jnp.repeat(x, 2)
Array([0, 0, 1, 1, 2, 2], dtype=int32)
>>> jnp.repeat(x, 2, total_repeat_length=x.size*2)  # the documented way to allow JIT-ing
Array([0, 0, 1, 1, 2, 2], dtype=int32)

>>> def func(x):
...     return jnp.repeat(x, 2)
... 
>>> func(x)
Array([0, 0, 1, 1, 2, 2], dtype=int32)
>>> # It's not actually needed to use `total_repeat_length` if `repeats` is a literal int:
>>> jax.jit(func)(x)
Array([0, 0, 1, 1, 2, 2], dtype=int32)

>>> # It is needed if we make `repeats` data-dependent:
>>> def func(x):
...     return jnp.repeat(x, x[2])
... 
>>> func(x)
Array([0, 0, 1, 1, 2, 2], dtype=int32)
>>> jax.jit(func)(x)
...
ConcretizationTypeError: Abstract tracer value encountered where concrete value is expected: Traced<ShapedArray(int32[])>with<DynamicJaxprTrace(level=0/1)>
```

A few conclusions:
- the main use case of this function, `repeats=a_literal_int`, is not data-dependent,
- the second most common use case of this function, `repeats=an_int_derived_from_x`, _is_ data-dependent,
- if we do want to include `repeats` as an array, that is almost always data-dependent
- Now that I've played with this some more: in case `repeats` is not an integer, it has to have the length of the number of elements of the first input array. The argument is inherently an array. Both JAX and PyTorch don't allow list/tuple, and while CuPy does allow (in fact, require) a sequence, its argument doesn't seem strong because the main use case requires `repeats=x.tolist()` (workaround from https://github.com/cupy/cupy/issues/3849) which gains nothing.
  - A hardcoded `repeat(x, [2, 3, 4, 5])` is highly unusual. That code know that `x` is of length 4 - may happen in tests, but is unlikely to be useful in production code.
  - This kind of usage is expected more (from SciPy code):

```python
# from scipy.signal tests
repeats[1::2] = x[1::2]
x = np.repeat(x, repeats)

# from scipy.integrate functionality
diff.data /= np.repeat(h, np.diff(diff.indptr))
```

My suggested resolution:
- do not allow sequence input for `repeats`, only integer or array
- do include the data-dependent note, with an explanation of when this occurs
- state that this API must always be implemented for the non-data-dependent part (`repeats` is a literal int), and only the data-dependent part is optional

kgryte commented:
@rgommers We discussed making `repeats` a sequence previously (result from that discussion: https://github.com/data-apis/array-api/issues/654#issuecomment-1769657879). In that thread, Mario considered it undesirable for `repeats` to be an array and would prefer PyTorch accept a list/sequence due to device synchronization.

rgommers commented:
> We discussed making `repeats` a sequence previously (result from that discussion: [#654 (comment)](https://github.com/data-apis/array-api/issues/654#issuecomment-1769657879)). In that thread, Mario considered it undesirable for `repeats` to be an array and would prefer PyTorch accept a list/sequence due to device synchronization.

I'd say I agree with Leo's comments in that thread. There's just not much of a point of it being a sequence. It is not like you can do `repeat(x, [2, 3])` and have that mean some kind of generalization of `repeat(x, 2)`. The `2` here is not an axis/shape, it's shorthand for `repeat(x, [2] * x.size)`.

NumPy also documents it as an ndarray; the only reason a sequence works for NumPy is because it calls `asarray` on all its array-like inputs.

leofang commented:
Q: Does it make sense to say "when the input is an array, this API has data-dependent output shape" and followed by the note?

kgryte commented:
Okay. I've dropped support for sequences and updated the admonition. The admonition now only allows optional support for providing an array as the second argument; all conforming libraries must support providing an integer.

As providing sequences is still controversial and can be added in a subsequent revision of the standard, I think it is fine to omit for the time being.

rgommers reviewed (commented):
Thanks @kgryte. The `repeats` treatment LGTM. Two other minor comments.

kgryte commented on src/array_api_stubs/_draft/manipulation_functions.py:
```diff
@@ -159,6 +160,53 @@ def permute_dims(x: array, /, axes: Tuple[int, ...]) -> array:
+    Returns
+    -------
+    out: array
+        an output array containing repeated elements. The returned array must have the same data type as ``x``. If ``axis`` is ``None``, the returned array must be a one-dimensional array; otherwise, the returned array have the same shape as ``x``, except for the axis (dimension) along which elements were repeated.
```
Sorry for the typo. Fixed.

kgryte commented on src/array_api_stubs/_draft/manipulation_functions.py:
```diff
@@ -159,6 +160,53 @@ def permute_dims(x: array, /, axes: Tuple[int, ...]) -> array:
+    Parameters
+    ----------
+    x: array
+        input array containing elements to repeat.
```
I believe the alternative you describe is `tile`. Updated the main description to be more explicit.

rgommers reviewed (approved):
LGTM now. I think this should be good to go; will aim to merge this at the end of today unless there are new comments.

rgommers commented:
Thanks @kgryte & all reviewers!
